Add precision conversion functions

upsj opened the pull request:
This PR adds precision conversion support to the Ginkgo matrix formats

TODO:
- [x] Implement array conversions
- [x] Implement matrix conversions

upsj commented:
The PR is now ready for review. There is one rather difficult issue: `strategy_type` in `Csr` and `Hybrid` cannot be converted, so I built a very annoying if-else monster for `Csr` and left the strategy type out completely for `Hybrid`, because the `strategy_type` implementations hide their parameters.

yhmtsai commented on core/device_hooks/common_kernels.inc.cpp:
```diff
@@ -66,6 +68,22 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+GKO_NOT_COMPILED(GKO_HOOK_MODULE);
+GKO_INSTANTIATE_FOR_EACH_INDEX_TYPE(GKO_DECLARE_PREFIX_SUM_KERNEL);
+// explicitly instantiate for size_type, as this is used in the SellP format
+template GKO_DECLARE_PREFIX_SUM_KERNEL(size_type);
```
Should these be in a namespace?

upsj replied:
Yes, I'll probably move them into a `precision_conversion` namespace. Not sure about `prefix_sum` though. That one I forgot to add to the `common_kernels`, but the name is already kind of self-explanatory. Do we need to put it into a namespace?

pratikvn commented on core/matrix/csr.cpp:
```diff
@@ -120,6 +120,68 @@ void Csr<ValueType, IndexType>::apply_impl(const LinOp *alpha, const LinOp *b,
+    result->row_ptrs_ = this->row_ptrs_;
+    result->set_size(this->get_size());
+    auto strat = this->get_strategy().get();
+    using Other = Csr<next_precision<ValueType>, IndexType>;
```
Ah I get it now, forget I asked this question.

upsj commented on core/base/array.cpp:
```diff
@@ -0,0 +1,71 @@
+
+
+namespace gko {
+namespace conversion {
```
Since it's only for the kernel operation, I think that's fine!

thoasm commented on include/ginkgo/core/base/array.hpp:
```diff
@@ -366,6 +383,54 @@ class Array {
+     * @return this
+     */
+    template <typename OtherValueType>
+    xstd::enable_if_t<!std::is_same<ValueType, OtherValueType>::value, Array>
+        &operator=(const Array<OtherValueType> &other)
```
Could you please also add a `operator=(Array<OtyerValueType> &&)` version?

upsj replied:
Can you give a use-case for this operator (i.e. where providing a second overload gives us an advantage?) or how the move-assignment should differ from copy-assignment?

thoasm replied:
With this operator I wasn't sure myself. At the time I wrote the comment, I simply thought: We support a `move` where we also support `operator=`, so the `move` is missing here.
I am actually fine if we skip the `move` here because we can never really reuse anything, so moving would not have any advantage over the copying.

thoasm commented on include/ginkgo/core/base/math.hpp:
```diff
@@ -202,6 +202,26 @@ GKO_INLINE GKO_ATTRIBUTES constexpr bool is_complex()
 namespace detail {
 
 
+// singly linked list of all our supported precisions
+template <typename T>
+struct next_precision_impl {};
+
+template <>
+struct next_precision_impl<float> {
+    using type = double;
+};
+
+template <>
+struct next_precision_impl<double> {
+    using type = float;
+};
+
+template <typename T>
+struct next_precision_impl<std::complex<T>> {
+    using type = std::complex<typename next_precision_impl<T>::type>;
+};
```
The problem I see with this is that we could no longer use types like `floatX` as `ValueType`, which worked before (I am not 100% sure if we advertised that, but we aimed to support it, see #49 and #64).
Is it possible to add a default dummy-type, so everything would still compile / work?

Alternatively, we could add a helper `supports_precision_change`, which would work in combination with [`std::conditional`](https://en.cppreference.com/w/cpp/types/conditional) for the inheritance and by its own with the `Array` class.

upsj replied:
We could add new types by adding them to the `next_precision` chain:
Instead of `float -> double -> float`, we could have `float -> double -> half -> bfloat16 -> unum -> double` or add multiple `next_x_precision` for larger steps in the chain (this would allow all pair-wise conversions directly)

thoasm replied:
I don't necessarily want something truly genetic, but I would like a `Dense` matrix to still compile when we use `floatx` as a type instead of our defaults. Obviously, we could not use the precision change feature in that case (which is totally reasonable), but it would be nice if we could use all the other features that do not depend on the precision change.

upsj replied:
Can you clarify how this would work together with our header/source separation? To me it looks like adding a new precision type requires changes at the implementation level, since it also requires modifying `dense.cpp` or re-implementing all member functions you want to use by yourself.

thoasm replied:
Oh, my bad, I thought we could simply use Ginkgo inside CMake, and use a `floatx` type as a template parameter. After looking into the PRs that I referenced, we simply added the new type to the instantiate macro to make it work. In this case, we would also have to add it to the `next_precision` implementation, but that seems fine to me.
Sorry for causing this much confusion.

thoasm commented on include/ginkgo/core/matrix/dense.hpp:
```diff
@@ -35,6 +35,7 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 
 #include <initializer_list>
+#include <type_traits>
```
nit: Where do you use components from `type_traits`?
You are missing the header `ginkgo/core/base/math.hpp`, which contains the `next_precision` type.
Pretty much all of the matrix formats are missing the inclusion of that header.

upsj replied:
That's leftover from when I was using `std::conditional` instead of `next_precision`. Good point!

thoasm commented on omp/test/components/precision_conversion.cpp:
```diff
@@ -0,0 +1,147 @@
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/components/prefix_sum.hpp"
```
Did you mean:
```suggestion
#include "core/components/precision_conversion.hpp"
```
Which you are actually not using, so can this header be removed?

upsj replied:
Copy-Paste... Yes, that shouldn't be there

upsj commented:
@yhmtsai float -> double is supposed to be deterministic, double -> float might depend on the rounding mode (which might depend on the issued instructions, flags set in the CPU, ...), i.e. rounding up, down, towards zero, nearest even/odd...

On the suggested tests, I need some clarification on how they extend what is already there

1. can you elaborate what you mean by *corresponding float* value for double -> float? And do you mean take double values that are not *exactly* representable as float (missing precision) or values that are not *at all* representable (outside value range)?

2. This seems to me like a subset of testing float -> double -> float for equality, since all of these double values are exactly representable as float

Maybe give a simple example with a single value?

thoasm commented on reference/test/components/precision_conversion.cpp:
```diff
@@ -0,0 +1,127 @@
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/components/prefix_sum.hpp"
```
Same with this header. Can it be removed?

thoasm commented on reference/test/matrix/coo_kernels.cpp:
```diff
@@ -98,6 +98,48 @@ class Coo : public ::testing::Test {
+    // If OtherType is more precise: 0, otherwise r
+    auto residual = r<OtherType>::value < r<ValueType>::value
+                        ? gko::remove_complex<ValueType>{0}
+                        : gko::remove_complex<OtherType>{r<OtherType>::value};
```
nit: The `r` struct already removes complex types (see definition [here](https://github.com/ginkgo-project/ginkgo/blob/develop/core/test/utils.hpp#L90))
```suggestion
                        : r<OtherType>::value;
```

upsj replied:
The point of these lines was actually to make sure that both branches of the ternary operator have the same type, so this was supposed to be
```suggestion
                        : gko::remove_complex<ValueType>{r<OtherType>::value};
```
Do you think that's okay? Or overly cautious?

thoasm replied:
That is fine for me as well.

thoasm commented on reference/test/matrix/coo_kernels.cpp:
```diff
@@ -98,6 +98,48 @@ class Coo : public ::testing::Test {
+    // If OtherType is more precise: 0, otherwise r
+    auto residual = r<OtherType>::value < r<ValueType>::value
+                        ? gko::remove_complex<ValueType>{0}
+                        : gko::remove_complex<OtherType>{r<OtherType>::value};
```
nit: same here
```suggestion
                        : r<OtherType>::value;
```

thoasm commented on reference/test/matrix/csr_kernels.cpp:
```diff
@@ -601,6 +601,54 @@ TYPED_TEST(Csr, ApplyFailsOnWrongNumberOfCols)
+    // If OtherType is more precise: 0, otherwise r
+    auto residual = r<OtherType>::value < r<ValueType>::value
+                        ? gko::remove_complex<ValueType>{0}
+                        : gko::remove_complex<OtherType>{r<OtherType>::value};
```
```suggestion
                        : r<OtherType>::value;
```

thoasm commented on reference/test/matrix/csr_kernels.cpp:
```diff
@@ -601,6 +601,54 @@ TYPED_TEST(Csr, ApplyFailsOnWrongNumberOfCols)
+    // If OtherType is more precise: 0, otherwise r
+    auto residual = r<OtherType>::value < r<ValueType>::value
+                        ? gko::remove_complex<ValueType>{0}
+                        : gko::remove_complex<OtherType>{r<OtherType>::value};
```
```suggestion
                        : r<OtherType>::value;
```

thoasm commented on reference/test/matrix/dense_kernels.cpp:
```diff
@@ -308,6 +308,46 @@ TYPED_TEST(Dense, ComputDotFailsOnWrongResultSize)
+    // If OtherT is more precise: 0, otherwise r
+    auto residual = r<OtherT>::value < r<T>::value
+                        ? gko::remove_complex<T>{0}
+                        : gko::remove_complex<T>{r<OtherT>::value};
```
```suggestion
                        : r<OtherT>::value;
```

thoasm commented on reference/test/matrix/dense_kernels.cpp:
```diff
@@ -308,6 +308,46 @@ TYPED_TEST(Dense, ComputDotFailsOnWrongResultSize)
+    // If OtherT is more precise: 0, otherwise r
+    auto residual = r<OtherT>::value < r<T>::value
+                        ? gko::remove_complex<T>{0}
+                        : gko::remove_complex<T>{r<OtherT>::value};
```
```suggestion
                        : r<OtherT>::value;
```

thoasm commented on reference/test/matrix/ell_kernels.cpp:
```diff
@@ -211,6 +211,48 @@ TYPED_TEST(Ell, ApplyFailsOnWrongNumberOfCols)
+    // If OtherType is more precise: 0, otherwise r
+    auto residual = r<OtherType>::value < r<ValueType>::value
+                        ? gko::remove_complex<ValueType>{0}
+                        : gko::remove_complex<OtherType>{r<OtherType>::value};
```
```suggestion
                        : r<OtherType>::value;
```

thoasm commented on reference/test/matrix/ell_kernels.cpp:
```diff
@@ -211,6 +211,48 @@ TYPED_TEST(Ell, ApplyFailsOnWrongNumberOfCols)
+    // If OtherType is more precise: 0, otherwise r
+    auto residual = r<OtherType>::value < r<ValueType>::value
+                        ? gko::remove_complex<ValueType>{0}
+                        : gko::remove_complex<OtherType>{r<OtherType>::value};
```
```suggestion
                        : r<OtherType>::value;
```

thoasm commented on reference/test/matrix/hybrid_kernels.cpp:
```diff
@@ -237,6 +237,48 @@ TYPED_TEST(Hybrid, ApplyFailsOnWrongNumberOfCols)
+    // If OtherType is more precise: 0, otherwise r
+    auto residual = r<OtherType>::value < r<ValueType>::value
+                        ? gko::remove_complex<ValueType>{0}
+                        : gko::remove_complex<OtherType>{r<OtherType>::value};
```
```suggestion
                        : r<OtherType>::value;
```

thoasm commented on reference/test/matrix/hybrid_kernels.cpp:
```diff
@@ -237,6 +237,48 @@ TYPED_TEST(Hybrid, ApplyFailsOnWrongNumberOfCols)
+    // If OtherType is more precise: 0, otherwise r
+    auto residual = r<OtherType>::value < r<ValueType>::value
+                        ? gko::remove_complex<ValueType>{0}
+                        : gko::remove_complex<OtherType>{r<OtherType>::value};
```
```suggestion
                        : r<OtherType>::value;
```

thoasm commented on reference/test/matrix/sellp_kernels.cpp:
```diff
@@ -187,6 +187,48 @@ TYPED_TEST(Sellp, ApplyFailsOnWrongNumberOfCols)
+    // If OtherType is more precise: 0, otherwise r
+    auto residual = r<OtherType>::value < r<ValueType>::value
+                        ? gko::remove_complex<ValueType>{0}
+                        : gko::remove_complex<OtherType>{r<OtherType>::value};
```
```suggestion
                        : r<OtherType>::value;
```

thoasm commented on reference/test/matrix/sellp_kernels.cpp:
```diff
@@ -187,6 +187,48 @@ TYPED_TEST(Sellp, ApplyFailsOnWrongNumberOfCols)
+    // If OtherType is more precise: 0, otherwise r
+    auto residual = r<OtherType>::value < r<ValueType>::value
+                        ? gko::remove_complex<ValueType>{0}
+                        : gko::remove_complex<OtherType>{r<OtherType>::value};
```
```suggestion
                        : r<OtherType>::value;
```

upsj commented:
@thoasm Thanks for your feedback! I would consider all of the `move_to` implementations something we should tackle in another PR, since as far as I can see, we currently don't really move anywhere (except for moves to the same type), and since our matrix classes have lots of assumptions/invariants about the stored data, we need to be very careful not to leave a matrix in an invalid state after a move (like a naive Coo -> Csr move would leave the row_idxs array unchanged, which gives inconsistent element numbers, a Csr -> * move must restore the row_ptrs or set the size to 0, ...)

yhmtsai commented:
@upsj I would like to see losing precision from these tests.
We can get the same result from float->float->float as float -> double -> float although I know it should be stored in double in the middle. These tests check whether float->double or double->float really changes the datatypes.
If double -> float is not deterministic, I think checking double->float(->double) with losing precision is enough. 

use double as X=(0x4218888000889111), float Y=(0x50c44400) and the double Z=(0x4218888000000000)?
Y=float(X), Z=double(Y) in Matlab.
Matlab turns this double value to the float value.
I get the error 1.2961e-09 from `abs(double(Y)-X)/X`, which is similar with the matrix test you did.

1. Doing the float->double on Y and comparing with Z.
2. Doing the double->float->double on X and comparing with X.

upsj commented:
I added tests for conversions via another executor (and also found a bug that way), rounding with precision loss and moved the component kernels to a namespace.

Thanks for your feedback, that was really helpful!

thoasm commented on cuda/components/precision_conversion.cu:
```diff
@@ -0,0 +1,67 @@
+#include "common/components/precision_conversion.hpp.inc"
+
+
+constexpr int default_block_size = 512;
```
nit: We usually have this defined before we include the `.inc` file, so we have only one defined variable `default_block_size` that we use in both files (cuda & components or hip & components).

thoasm commented on hip/components/precision_conversion.hip.cpp:
```diff
@@ -0,0 +1,68 @@
+#include "common/components/precision_conversion.hpp.inc"
+
+
+constexpr int default_block_size = 512;
```
nit: see previous comment on moving this above the include to avoid having 2 `default_block_size`.

thoasm commented on cuda/test/components/precision_conversion.cu:
```diff
@@ -0,0 +1,173 @@
+        }
+        dvals = vals;
+        dcvals = cvals;
+        gko::uint64 rawdouble = 0x4218888000889111ULL;
+        gko::uint32 rawfloat = 0x50c44400ULL;
+        gko::uint64 rawrounded = 0x4218888000000000ULL;
```
On all of these, I would prefer the brace-initializer `{}` to make surer it can fit.
For `rawfloat`, your suffix is actually wrong (ULL is 64 bit):
```suggestion
        gko::uint64 rawdouble{0x4218888000889111ULL};
        gko::uint32 rawfloat{0x50c44400U};
        gko::uint64 rawrounded{0x4218888000000000ULL};
```
I think it should also be possible to remove all suffixes without any issues, but since I did not test it, I left them in.

upsj replied:
Small detail: `U` only guarantees 16 bit, it needs to be `UL`

thoasm replied:
I don't think so, since the number that you wrote can't be stored in 16 bits, so C++ chooses the next larger type that can store the number.
Even for the others, I don't think it is necessary to have any postfix, but they don't hurt as long as they have the appropriate size (which I would like to have checked with the brace-initialization).

upsj replied:
That is true. I went with braces in the final version ;)
Sometimes I like to be excessively verbose just to be sure :D

thoasm commented on hip/test/components/precision_conversion.hip.cpp:
```diff
@@ -0,0 +1,173 @@
+        }
+        dvals = vals;
+        dcvals = cvals;
+        gko::uint64 rawdouble = 0x4218888000889111ULL;
+        gko::uint32 rawfloat = 0x50c44400ULL;
+        gko::uint64 rawrounded = 0x4218888000000000ULL;
```
```suggestion
        gko::uint64 rawdouble{0x4218888000889111ULL};
        gko::uint32 rawfloat{0x50c44400U};
        gko::uint64 rawrounded{0x4218888000000000ULL};
```

thoasm commented on omp/test/components/precision_conversion.cpp:
```diff
@@ -0,0 +1,173 @@
+        }
+        dvals = vals;
+        dcvals = cvals;
+        gko::uint64 rawdouble = 0x4218888000889111ULL;
+        gko::uint32 rawfloat = 0x50c44400ULL;
+        gko::uint64 rawrounded = 0x4218888000000000ULL;
```
```suggestion
        gko::uint64 rawdouble{0x4218888000889111ULL};
        gko::uint32 rawfloat{0x50c44400U};
        gko::uint64 rawrounded{0x4218888000000000ULL};
```

thoasm commented on reference/test/components/precision_conversion.cpp:
```diff
@@ -0,0 +1,150 @@
+            vals.get_data()[i] = dist(rand);
+            cvals.get_data()[i] = {dist(rand), dist(rand)};
+        }
+        gko::uint64 rawdouble = 0x4218888000889111ULL;
+        gko::uint32 rawfloat = 0x50c44400ULL;
+        gko::uint64 rawrounded = 0x4218888000000000ULL;
```
```suggestion
        gko::uint64 rawdouble{0x4218888000889111ULL};
        gko::uint32 rawfloat{0x50c44400U};
        gko::uint64 rawrounded{0x4218888000000000ULL};
```

upsj commented:
@yhmtsai The tests make sure that conversions between different executors work. This actually uncovered a bug in the Array conversion-assignment implementation.

hartwiganzt commented:
@upsj ready to be merged!
